File: backend/app/routers/admin/projects.py
from datetime import datetime
from fastapi import APIRouter, Depends, HTTPException
from bson import ObjectId
from app.database import get_database
from app.middleware.auth import require_admin
from app.schemas.project import (
    CreateProjectRequest,
    UpdateProjectRequest,
    DeliverProjectRequest,
    ProjectResponse,
)
from app.models.project import new_project
from app.models.user import new_user
from app.services.auth import hash_password, create_invite_token
from app.services.email import send_invite_email, send_gallery_link_email
from app.services.s3 import delete_prefix_from_s3
from app.utils.tokens import generate_share_token
from app.models.invoice import new_invoice

router = APIRouter()


def _project_response(p: dict) -> ProjectResponse:
    return ProjectResponse(
        id=str(p["_id"]),
        title=p["title"],
        description=p["description"],
        client_id=str(p["client_id"]),
        status=p["status"],
        cover_image_key=p.get("cover_image_key"),
        categories=p.get("categories", []),
        share_link_token=p.get("share_link_token"),
        share_link_expires_at=p.get("share_link_expires_at"),
        project_expires_at=p.get("project_expires_at"),
        created_at=p["created_at"],
        updated_at=p["updated_at"],
    )


@router.get("")
async def list_projects(status: str | None = None, client_id: str | None = None, admin: dict = Depends(require_admin)):
    db = get_database()
    query = {}
    if status:
        query["status"] = status
    if client_id:
        query["client_id"] = client_id
    cursor = db.projects.find(query).sort("created_at", -1)
    items = []
    async for p in cursor:
        items.append(_project_response(p))
    return {"projects": items}


@router.post("", status_code=201)
async def create_project(body: CreateProjectRequest, admin: dict = Depends(require_admin)):
    db = get_database()
    # Find or create client
    client = await db.users.find_one({"email": body.client_email})
    if not client:
        invite_token = create_invite_token()
        client_doc = new_user(
            email=body.client_email,
            password_hash=hash_password(invite_token),  # temporary
            role="client",
            name=body.client_name,
        )
        client_doc["invite_token"] = invite_token
        result = await db.users.insert_one(client_doc)
        client_id = str(result.inserted_id)
        send_invite_email(body.client_email, body.client_name, invite_token)
    else:
        client_id = str(client["_id"])
    project = new_project(
        title=body.title,
        client_id=client_id,
        description=body.description,
        categories=body.categories,
    )
    insert = await db.projects.insert_one(project)
    # Auto-book inquiry if linked
    if body.inquiry_id:
        await db.inquiries.update_one(
            {"_id": ObjectId(body.inquiry_id)},
            {"$set": {"status": "booked"}},
        )
    return {"id": str(insert.inserted_id), "client_id": client_id, "message": "Project created"}


@router.get("/{project_id}")
async def get_project(project_id: str, admin: dict = Depends(require_admin)):
    db = get_database()
    p = await db.projects.find_one({"_id": ObjectId(project_id)})
    if not p:
        raise HTTPException(status_code=404, detail="Project not found")
    return _project_response(p)


@router.put("/{project_id}")
async def update_project(project_id: str, body: UpdateProjectRequest, admin: dict = Depends(require_admin)):
    db = get_database()
    update = {k: v for k, v in body.model_dump(exclude_none=True).items()}
    if not update:
        raise HTTPException(status_code=400, detail="No fields to update")
    update["updated_at"] = datetime.utcnow()
    result = await db.projects.update_one({"_id": ObjectId(project_id)}, {"$set": update})
    if result.matched_count == 0:
        raise HTTPException(status_code=404, detail="Project not found")
    return {"message": "Updated"}


@router.delete("/{project_id}")
async def delete_project(project_id: str, admin: dict = Depends(require_admin)):
    db = get_database()
    project = await db.projects.find_one({"_id": ObjectId(project_id)})
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    delete_prefix_from_s3(f"projects/{project_id}/")
    await db.media.delete_many({"project_id": project_id})
    await db.projects.delete_one({"_id": ObjectId(project_id)})
    return {"message": "Project and all media deleted"}


@router.post("/{project_id}/deliver")
async def deliver_project(project_id: str, body: DeliverProjectRequest, admin: dict = Depends(require_admin)):
    db = get_database()
    project = await db.projects.find_one({"_id": ObjectId(project_id)})
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    token = generate_share_token()
    update = {
        "share_link_token": token,
        "status": "delivered",
        "updated_at": datetime.utcnow(),
    }
    if body.share_link_expires_at:
        update["share_link_expires_at"] = body.share_link_expires_at
    if body.project_expires_at:
        update["project_expires_at"] = body.project_expires_at
    await db.projects.update_one({"_id": ObjectId(project_id)}, {"$set": update})
    # Optionally create an invoice
    invoice_token = None
    if body.create_invoice:
        if not body.invoice_line_items or not body.invoice_due_date:
            raise HTTPException(status_code=400, detail="Line items and due date required when creating an invoice")
        total_cents = sum(li.amount_cents * li.quantity for li in body.invoice_line_items)
        line_items_dicts = [li.model_dump() for li in body.invoice_line_items]
        invoice = new_invoice(
            client_id=str(project["client_id"]),
            amount_cents=total_cents,
            line_items=line_items_dicts,
            due_date=body.invoice_due_date,
            project_id=project_id,
        )
        invoice["status"] = "sent"
        await db.invoices.insert_one(invoice)
        invoice_token = invoice["token"]
    # Notify client
    client = await db.users.find_one({"_id": ObjectId(project["client_id"])})
    if client:
        send_gallery_link_email(client["email"], client.get("name", ""), token, project["title"])
    return {"share_link_token": token, "invoice_token": invoice_token, "message": "Project delivered"}


@router.put("/{project_id}/rescind")
async def rescind_delivery(project_id: str, admin: dict = Depends(require_admin)):
    db = get_database()
    project = await db.projects.find_one({"_id": ObjectId(project_id)})
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    if project["status"] != "delivered":
        raise HTTPException(status_code=400, detail="Only delivered projects can be rescinded")
    await db.projects.update_one(
        {"_id": ObjectId(project_id)},
        {"$set": {"status": "active", "share_link_token": None, "share_link_expires_at": None, "updated_at": datetime.utcnow()}},
    )
    return {"message": "Delivery rescinded, project returned to active"}


@router.put("/{project_id}/archive")
async def archive_project(project_id: str, admin: dict = Depends(require_admin)):
    db = get_database()
    project = await db.projects.find_one({"_id": ObjectId(project_id)})
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    delete_prefix_from_s3(f"projects/{project_id}/")
    await db.media.delete_many({"project_id": project_id})
    await db.projects.update_one(
        {"_id": ObjectId(project_id)},
        {"$set": {"status": "archived", "updated_at": datetime.utcnow()}},
    )
    return {"message": "Project archived"}

File: src/routes/admin/projects.ts
import { Router, Response } from 'express';
import { ObjectId } from 'mongodb';
import { getDb } from '../../database';
import { requireAdmin, AuthRequest } from '../../middleware/auth';
import { newProject } from '../../models/project';
import { newUser } from '../../models/user';
import { newInvoice } from '../../models/invoice';
import { hashPassword, createInviteToken } from '../../services/auth';
import { sendInviteEmail, sendGalleryLinkEmail } from '../../services/email';
import { deletePrefixFromS3 } from '../../services/s3';
import { generateShareToken } from '../../utils/tokens';

const router = Router();

function projectResponse(p: Record<string, unknown>) {
  return {
    id: (p._id as ObjectId).toString(),
    title: p.title,
    description: p.description,
    client_id: p.client_id,
    status: p.status,
    cover_image_key: p.cover_image_key ?? null,
    categories: p.categories ?? [],
    share_link_token: p.share_link_token ?? null,
    share_link_expires_at: p.share_link_expires_at ?? null,
    project_expires_at: p.project_expires_at ?? null,
    created_at: p.created_at,
    updated_at: p.updated_at,
  };
}

router.get('/', requireAdmin, async (req: AuthRequest, res: Response): Promise<void> => {
  const db = await getDb();
  const query: Record<string, unknown> = {};
  if (req.query.status) query.status = req.query.status;
  if (req.query.client_id) query.client_id = req.query.client_id;
  const projects = await db.collection('projects').find(query).sort({ created_at: -1 }).toArray();
  res.json({ projects: projects.map(projectResponse) });
});

router.post('/', requireAdmin, async (req: AuthRequest, res: Response): Promise<void> => {
  const { title, client_email, client_name, description, categories, inquiry_id } = req.body;
  if (!title || !client_email || !client_name) { res.status(400).json({ detail: 'title, client_email, client_name required' }); return; }
  const db = await getDb();
  let clientId: string;
  const existing = await db.collection('users').findOne({ email: client_email });
  if (!existing) {
    const inviteToken = createInviteToken();
    const clientDoc = newUser(client_email, hashPassword(inviteToken), 'client', client_name);
    (clientDoc as Record<string, unknown>).invite_token = inviteToken;
    const result = await db.collection('users').insertOne(clientDoc);
    clientId = result.insertedId.toString();
    await sendInviteEmail(client_email, client_name, inviteToken);
  } else {
    clientId = existing._id.toString();
  }
  const project = newProject(title, clientId, description ?? '', categories ?? []);
  const insert = await db.collection('projects').insertOne(project);
  if (inquiry_id) {
    await db.collection('inquiries').updateOne({ _id: new ObjectId(inquiry_id) }, { $set: { status: 'booked' } });
  }
  res.status(201).json({ id: insert.insertedId.toString(), client_id: clientId, message: 'Project created' });
});

router.get('/:projectId', requireAdmin, async (req: AuthRequest, res: Response): Promise<void> => {
  const db = await getDb();
  const p = await db.collection('projects').findOne({ _id: new ObjectId(req.params.projectId) });
  if (!p) { res.status(404).json({ detail: 'Project not found' }); return; }
  res.json(projectResponse(p));
});

router.put('/:projectId', requireAdmin, async (req: AuthRequest, res: Response): Promise<void> => {
  const allowed = ['title', 'description', 'categories', 'cover_image_key', 'status', 'share_link_expires_at', 'project_expires_at'];
  const update: Record<string, unknown> = {};
  for (const key of allowed) {
    if (req.body[key] !== undefined) update[key] = req.body[key];
  }
  if (!Object.keys(update).length) { res.status(400).json({ detail: 'No fields to update' }); return; }
  update.updated_at = new Date();
  const db = await getDb();
  const result = await db.collection('projects').updateOne({ _id: new ObjectId(req.params.projectId) }, { $set: update });
  if (!result.matchedCount) { res.status(404).json({ detail: 'Project not found' }); return; }
  res.json({ message: 'Updated' });
});

router.delete('/:projectId', requireAdmin, async (req: AuthRequest, res: Response): Promise<void> => {
  const db = await getDb();
  const project = await db.collection('projects').findOne({ _id: new ObjectId(req.params.projectId) });
  if (!project) { res.status(404).json({ detail: 'Project not found' }); return; }
  await deletePrefixFromS3(`projects/${req.params.projectId}/`);
  await db.collection('media').deleteMany({ project_id: req.params.projectId });
  await db.collection('projects').deleteOne({ _id: new ObjectId(req.params.projectId) });
  res.json({ message: 'Project and all media deleted' });
});

router.post('/:projectId/deliver', requireAdmin, async (req: AuthRequest, res: Response): Promise<void> => {
  const db = await getDb();
  const project = await db.collection('projects').findOne({ _id: new ObjectId(req.params.projectId) });
  if (!project) { res.status(404).json({ detail: 'Project not found' }); return; }
  const token = generateShareToken();
  const update: Record<string, unknown> = { share_link_token: token, status: 'delivered', updated_at: new Date() };
  if (req.body.share_link_expires_at) update.share_link_expires_at = new Date(req.body.share_link_expires_at);
  if (req.body.project_expires_at) update.project_expires_at = new Date(req.body.project_expires_at);
  await db.collection('projects').updateOne({ _id: new ObjectId(req.params.projectId) }, { $set: update });

  let invoiceToken: string | null = null;
  if (req.body.create_invoice) {
    const { invoice_line_items, invoice_due_date } = req.body;
    if (!invoice_line_items || !invoice_due_date) {
      res.status(400).json({ detail: 'Line items and due date required when creating an invoice' }); return;
    }
    const totalCents = invoice_line_items.reduce((sum: number, li: { amount_cents: number; quantity: number }) => sum + li.amount_cents * li.quantity, 0);
    const invoice = newInvoice(project.client_id as string, totalCents, invoice_line_items, new Date(invoice_due_date), req.params.projectId);
    (invoice as Record<string, unknown>).status = 'sent';
    await db.collection('invoices').insertOne(invoice);
    invoiceToken = (invoice as Record<string, unknown>).token as string;
  }

  const client = await db.collection('users').findOne({ _id: new ObjectId(project.client_id as string) });
  if (client) {
    await sendGalleryLinkEmail(client.email, client.name ?? '', token, project.title as string);
  }
  res.json({ share_link_token: token, invoice_token: invoiceToken, message: 'Project delivered' });
});

router.put('/:projectId/rescind', requireAdmin, async (req: AuthRequest, res: Response): Promise<void> => {
  const db = await getDb();
  const project = await db.collection('projects').findOne({ _id: new ObjectId(req.params.projectId) });
  if (!project) { res.status(404).json({ detail: 'Project not found' }); return; }
  if (project.status !== 'delivered') { res.status(400).json({ detail: 'Only delivered projects can be rescinded' }); return; }
  await db.collection('projects').updateOne(
    { _id: new ObjectId(req.params.projectId) },
    { $set: { status: 'active', share_link_token: null, share_link_expires_at: null, updated_at: new Date() } },
  );
  res.json({ message: 'Delivery rescinded, project returned to active' });
});

router.put('/:projectId/archive', requireAdmin, async (req: AuthRequest, res: Response): Promise<void> => {
  const db = await getDb();
  const project = await db.collection('projects').findOne({ _id: new ObjectId(req.params.projectId) });
  if (!project) { res.status(404).json({ detail: 'Project not found' }); return; }
  await deletePrefixFromS3(`projects/${req.params.projectId}/`);
  await db.collection('media').deleteMany({ project_id: req.params.projectId });
  await db.collection('projects').updateOne({ _id: new ObjectId(req.params.projectId) }, { $set: { status: 'archived', updated_at: new Date() } });
  res.json({ message: 'Project archived' });
});

export default router;
